File: impacted-test-engine/src/main/java/com/teamscale/test_impacted/engine/executor/ImpactedTestsSorter.java
package com.teamscale.test_impacted.engine.executor;

import com.teamscale.client.PrioritizableTest;
import com.teamscale.client.PrioritizableTestCluster;
import com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils;
import org.junit.platform.engine.TestDescriptor;
import org.junit.platform.engine.UniqueId;

import java.util.ArrayList;
import java.util.Collections;
import java.util.IdentityHashMap;
import java.util.List;
import java.util.Optional;
import java.util.Set;

import static com.teamscale.test_impacted.engine.ImpactedTestEngine.LOGGER;

/**
 * Test sorter that requests impacted tests from Teamscale and rewrites the {@link TestDescriptor} to take the returned
 * order into account when executing the tests.
 */
public class ImpactedTestsSorter implements ITestSorter {

	private final ImpactedTestsProvider impactedTestsProvider;

	public ImpactedTestsSorter(ImpactedTestsProvider impactedTestsProvider) {
		this.impactedTestsProvider = impactedTestsProvider;
	}

	@Override
	public void selectAndSort(TestDescriptor rootTestDescriptor) {
		AvailableTests availableTests = TestDescriptorUtils
				.getAvailableTests(rootTestDescriptor, impactedTestsProvider.partition);

		List<PrioritizableTestCluster> testClusters = impactedTestsProvider.getImpactedTestsFromTeamscale(
				availableTests.getTestList());

		if (testClusters == null) {
			LOGGER.fine(() -> "Falling back to execute all!");
			return;
		}

		Set<? super TestDescriptor> testRepresentatives = Collections.newSetFromMap(new IdentityHashMap<>());
		Set<? super TestDescriptor> seenDescriptors = Collections.newSetFromMap(new IdentityHashMap<>());
		for (PrioritizableTestCluster testCluster : testClusters) {
			for (PrioritizableTest test : testCluster.tests) {
				Optional<UniqueId> uniqueId = availableTests.convertToUniqueId(test);
				if (!uniqueId.isPresent()) {
					LOGGER.severe(() -> "Falling back to execute all...");
					return;
				}
				Optional<? extends TestDescriptor> testDescriptor = rootTestDescriptor.findByUniqueId(uniqueId.get());
				if (!testDescriptor.isPresent()) {
					LOGGER.severe(() -> "Falling back to execute all...");
					return;
				}
				testRepresentatives.add(testDescriptor.get());
				reinsertIntoHierarchy(testDescriptor.get(), seenDescriptors);
			}
		}

		removeNonImpactedTests(rootTestDescriptor, testRepresentatives);
	}

	/**
	 * Reinserts the given testDescriptor into the hierarchy by walking up the parents chain. By doing this in order
	 * with all tests we end up with our intended order. This is continued until we reach a node that has already been
	 * reinserted in a previous run, because parents should be sorted according to the order of their most important
	 * child descriptors.
	 */
	private static void reinsertIntoHierarchy(TestDescriptor testDescriptor,
											  Set<? super TestDescriptor> seenDescriptors) {
		Optional<? extends TestDescriptor> currentTestDescriptor = Optional.of(testDescriptor);
		while (currentTestDescriptor.isPresent() && !seenDescriptors.contains(currentTestDescriptor.get())) {
			seenDescriptors.add(currentTestDescriptor.get());
			TestDescriptorUtils.reinsertIntoParent(currentTestDescriptor.get());
			currentTestDescriptor = currentTestDescriptor.get().getParent();
		}
	}

	private void removeNonImpactedTests(TestDescriptor testDescriptor, Set<? super TestDescriptor> testDescriptors) {
		if (testDescriptors.contains(testDescriptor)) {
			return;
		}
		for (TestDescriptor descriptor : new ArrayList<>(testDescriptor.getChildren())) {
			removeNonImpactedTests(descriptor, testDescriptors);
		}
		if (testDescriptor.getChildren().isEmpty()) {
			testDescriptor.removeFromHierarchy();
		}
	}
}

File: impacted-test-engine/src/main/kotlin/com/teamscale/test_impacted/engine/executor/ImpactedTestsSorter.kt
package com.teamscale.test_impacted.engine.executor

import com.teamscale.test_impacted.engine.ImpactedTestEngine
import com.teamscale.test_impacted.test_descriptor.TestDescriptorUtils.getAvailableTests
import org.junit.platform.engine.TestDescriptor
import java.util.*

/**
* Test sorter that requests impacted tests from Teamscale and rewrites the [TestDescriptor] to take the returned
* order into account when executing the tests.
*/
class ImpactedTestsSorter(private val impactedTestsProvider: ImpactedTestsProvider) : ITestSorter {

	override fun selectAndSort(testDescriptor: TestDescriptor) {
		val availableTests = getAvailableTests(testDescriptor, impactedTestsProvider.partition)

		val testClusters = impactedTestsProvider.getImpactedTestsFromTeamscale(availableTests.testList)

		if (testClusters == null) {
			ImpactedTestEngine.LOG.fine { "Falling back to execute all!" }
			return
		}

		val testRepresentatives = Collections.newSetFromMap<TestDescriptor>(IdentityHashMap())
		val seenDescriptors = Collections.newSetFromMap<TestDescriptor>(IdentityHashMap())

		val allTests = testClusters.asSequence()
			.flatMap { it.tests?.asSequence() ?: emptySequence() }

		allTests.forEach { test ->
			val uniqueId = availableTests.convertToUniqueId(test)
			if (!uniqueId.isPresent) {
				ImpactedTestEngine.LOG.severe { "Falling back to execute all..." }
				return
			}
			val availableTest = testDescriptor.findByUniqueId(uniqueId.get())
			if (!availableTest.isPresent) {
				ImpactedTestEngine.LOG.severe { "Falling back to execute all..." }
				return
			}
			val descriptor = availableTest.get()
			testRepresentatives.add(descriptor)
			reinsertIntoHierarchy(descriptor, seenDescriptors)
		}

		removeNonImpactedTests(testDescriptor, testRepresentatives)
	}

	private fun removeNonImpactedTests(
		testDescriptor: TestDescriptor,
		impactedTestDescriptors: Set<TestDescriptor>
	) {
		if (testDescriptor in impactedTestDescriptors) return

		testDescriptor.children.toList().forEach { child ->
			removeNonImpactedTests(child, impactedTestDescriptors)
		}

		if (testDescriptor.children.isEmpty() && !testDescriptor.isRoot) {
			testDescriptor.removeFromHierarchy()
		}
	}

	/**
	 * Reinserts the given [testDescriptor] into the hierarchy by walking up the parent chain recursively.
	 * This ensures that parent descriptors are sorted according to the order of their most important child descriptors.
	 */
	private tailrec fun reinsertIntoHierarchy(
		testDescriptor: TestDescriptor,
		seenDescriptors: MutableSet<TestDescriptor>
	) {
		if (!seenDescriptors.add(testDescriptor)) return
		testDescriptor.reinsertIntoParent()
		val parentDescriptor = testDescriptor.parent.orElse(null)
		if (parentDescriptor != null) {
			reinsertIntoHierarchy(parentDescriptor, seenDescriptors)
		}
	}

	/**
	 * Removes the [this@reinsertIntoParent] from its parent and re-inserts it.
	 * This moves the descriptor to the end of the iteration order in the parent's children,
	 * effectively reordering it.
	 */
	private fun TestDescriptor.reinsertIntoParent() {
		parent.ifPresent { parent ->
			parent.removeChild(this)
			parent.addChild(this)
		}
	}
}
